[Subtitles][WebVTT] Fixed segmented webvtt sync

CastagnaIT opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
This PR add a new member to the `DemuxPacket` to attach the pts correction value to the packet itself, this value will be used to adjust the webvtt sync.

The reason why i add this new member to the `DemuxPacket` instead of `DEMUX_PACKET` https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/inputstream/demux_packet.h#L45
is that `DemuxPacket` struct is used only internally the VP core, unlike `DEMUX_PACKET` is used on binary addons interface.

from my research this is the flow (correct me if i am wrong):
1) binary addon create `DEMUX_PACKET`
2) VP get the `DEMUX_PACKET` from binary addon and cast the packet as `DemuxPacket`
3) VP when needed apply the PTS correction by changing the value to the `DemuxPacket`
4) VP feed each demuxer (like subtitle webvtt) with this `DemuxPacket`

Since PTS correction value is set by VP (after get the `DEMUX_PACKET`) and there is no need to expose this new member to the addons interface, add the member for VP internal use only seem the way, also to avoid rebuild all binary addons for nothing,
to take in account that is needed to backport it to Kodi 20, the change is non-invasive

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
see: https://github.com/xbmc/inputstream.adaptive/pull/1082

there are many type of segmented webvtt cases for vod and live content,
that in the past we have tried to fix with PR #21523 but new use cases has raised other problems,
thanks to @glennguy we have solved most of the problem inside ISAdaptive addon,
but some of them require to adjust subtitle with the pts correction that is already included in the pts value itself, done by the VP player

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
These are cases but i cant attach here:
- segmented webvtt with chapters
- segmented webvtt live
- segmented webvtt VOD, D+ example, case that require pts correction to sync correctly
- fmp4 webvtt

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
Have segmented webvtt subtitles in-sync with the video

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CastagnaIT committed:
[Subtitles][WebVTT] Partial revert of 79719691a8208b061fe78309e88046a2be5167f6

CastagnaIT committed:
[Subtitles][WebVTT] Adjust offset by using pts correction

glennguy commented:
Just to give some more background:

Services like Disney+ often show a splash screen for some seconds before the actual program starts. Playlist-wise this is delivered in a different period/chapter, because the splash is a reused asset and not continuous with the program content. In other words, a 3 second splash screen runs from PTS 0 through to 3000000, then the new chapter starts and PTS is back at 0.

VP detects this discontinuity in `CVideoPlayer::CheckContinuity` and stores the difference in `m_offset_pts` which it applies to each incoming demux packet before passing it onto the respective VP subcomponent.

This isn't a problem with audio/video packets as the data within contains its own internal PTS and everything stays aligned. In the case with webvtt/hls subtitles though there is no internal PTS. It's just a plain text file with subtitle cues and a mapping of wall clock time to video/audio pts. We've done what we can in inputstream.adaptive to provide a PTS/DTS for subtitle packets that is aligned with the audio/video ones but cannot control or reliably predict what VP will do with applying this offset.

The changes that Stefano has put forward here 'preserves' the information of what offset was applied so we can then undo it once the subtitle packet is ready to be processed.

@fernetmenta - we would be grateful to get your thoughts on this. I hope I've described the current behaviour correctly, and if so is this the intended behaviour when starting a new chapter? In other words are we solving the problem correctly (from VP point of view) and not putting a band-aid on a different issue?

CastagnaIT committed:
[InputStreamAddon] Use reinterpret_cast to cast DemuxPacket

the use of static_cast do not allow the debugger to show struct members of DemuxPacket when you debugging

phunkyfish reviewed (approved):
Nice alternative to original PR, it's unobtrusive and does not require rebuilding of video inputstream add-ons either.
